core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Workflow.scala

```diff
@@ -150,177 +55,166 @@ class Workflow(
-      }
-    }
-    result
-  }
+//  def getSources(operator: OperatorIdentity): Set[OperatorIdentity] = {
+//    var result = Set[OperatorIdentity]()
+//    var current = Set[OperatorIdentity](operator)
+//    while (current.nonEmpty) {
+//      var next = Set[OperatorIdentity]()
+//      for (i <- current) {
+//        if (inLinks.contains(i) && inLinks(i).nonEmpty) {
+//          next ++= inLinks(i)
+//        } else {
+//          result += i
+//        }
+//        current = next
+//      }
+//    }
+//    result
+//  }
 
   def getWorkflowStatus: Map[String, OperatorRuntimeStats] = {
-    operatorToOpExecConfig.map { op =>
-      (op._1.operator, op._2.getOperatorStatistics)
-    }.toMap
+    physicalPlan.operatorMap.map(op => (op._1.operator, op._2.getOperatorStatistics))
   }
 
-  def getStartOperators: Iterable[OpExecConfig] = sourceOperators.map(operatorToOpExecConfig(_))
+//  def getStartOperators: Iterable[OpExecConfig] = sourceOperators.map(operatorToOpExecConfig(_))
+//
+//  def getEndOperators: Iterable[OpExecConfig] = sinkOperators.map(operatorToOpExecConfig(_))
 
-  def getEndOperators: Iterable[OpExecConfig] = sinkOperators.map(operatorToOpExecConfig(_))
+//  def getOperator(opID: String): OpExecConfig =
+//    operatorToOpExecConfig(OperatorIdentity(workflowId.id, opID))
+//
+//  def getOperator(opID: OperatorIdentity): OpExecConfig = operatorToOpExecConfig(opID)
 
-  def getOperator(opID: String): OpExecConfig =
-    operatorToOpExecConfig(OperatorIdentity(workflowId.id, opID))
+//  def getDirectUpstreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
+//    inLinks.getOrElse(opID, Set())
+//
+//  def getDirectDownStreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
+//    outLinks.getOrElse(opID, Set())
 
-  def getOperator(opID: OperatorIdentity): OpExecConfig = operatorToOpExecConfig(opID)
+  def getAllOperators: Iterable[NewOpExecConfig] = physicalPlan.operators
 
-  def getDirectUpstreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
-    inLinks.getOrElse(opID, Set())
-
-  def getDirectDownStreamOperators(opID: OperatorIdentity): Iterable[OperatorIdentity] =
-    outLinks.getOrElse(opID, Set())
-
-  def getAllOperators: Iterable[OpExecConfig] = operatorToOpExecConfig.values
-
-  def getWorkerInfo(id: ActorVirtualIdentity): WorkerInfo = workerToLayer(id).workers(id)
+  def getWorkerInfo(id: ActorVirtualIdentity): WorkerInfo = workerToOpExecConfig(id).workers(id)
 
   /**
     * Returns the worker layer of the upstream operators that links to the `opId` operator's
     * worker layer.
     */
   def getUpStreamConnectedWorkerLayers(
-      opID: OperatorIdentity
-  ): mutable.HashMap[OperatorIdentity, WorkerLayer] = {
-    val upstreamOperatorToLayers = new mutable.HashMap[OperatorIdentity, WorkerLayer]()
-    getDirectUpstreamOperators(opID).map(uOpID =>
-      upstreamOperatorToLayers(uOpID) = getOperator(uOpID).topology.layers.last
-    )
+      opID: LayerIdentity
```

Comment:
What's the difference between `LayerIdentity` and `OperatorIdentity` after this PR? Just curious.

Reply:
I have modified the PR description to explain the differences. `LayerID` means a `PhysicalOperatorID`. We'll do some renaming in later PRs, otherwise this PR will have too many changes.